flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/PartitionStateChecker.java

```diff
@@ -0,0 +1,63 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.io.network.partition;
+
+import org.apache.flink.runtime.execution.ExecutionState;
+import org.apache.flink.runtime.jobgraph.IntermediateDataSetID;
+import org.apache.flink.types.Either;
+
+import java.util.concurrent.CompletableFuture;
+
+/**
+ * Queries execution states of partition producer and consumer, accepts actions of state check result.
+ */
+public interface PartitionStateChecker {
+	/**
+	 * Triggers check of the producer and consumer execution states.
+	 *
+	 * @param intermediateDataSetId ID of the parent intermediate data set.
+	 * @param resultPartitionId ID of the result partition to check. This
+	 * identifies the producing execution and partition.
+	 * @return a future of check result.
+	 */
+	CompletableFuture<? extends CheckResult> triggerPartitionProducerCheck(
+		IntermediateDataSetID intermediateDataSetId,
+		ResultPartitionID resultPartitionId);
+
+	/**
+	 * Result of partition state check, accepts check callbacks.
+	 */
+	interface CheckResult {
+		ExecutionState getConsumerExecutionState();
```

Comment:
Do we really need the consumer's execution state here?

Reply:
At the moment we use it as a shortcut if consumer is done before the check has been accomplished then we skip the check. Ideally we should use closeFuture in SingleInputGate instead of ConsumerExecutionState for this. Although this shortcut is the best effort because the state change or gate close can happen concurrently right after this check, I suggest we consider it as a separate refactoring as it would change a bit current behaviour: https://issues.apache.org/jira/browse/FLINK-12672